### PR TITLE
(feat): add tx filterer contract

### PR DESCRIPTION
**Pull request by Raid5594**
# What ❔

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

### CODE REVIEW

File: l1-contracts/contracts/transactionFilterer/TransactionFilterer.sol [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+import {ReentrancyGuard} from "../common/ReentrancyGuard.sol";
+import {AlreadyWhitelisted, NotWhitelisted, ZeroAddress} from "../common/L1ContractErrors.sol";
+import {ITransactionFilterer} from "../state-transition/chain-interfaces/ITransactionFilterer.sol";
+
+/// @author Matter Labs
+/// @custom:security-contact security@matterlabs.dev
+/// @dev Filters transactions received by the Mailbox
+/// @dev Only allows whitelisted senders to deposit to Gateway
+contract TransactionFilterer is ITransactionFilterer, ReentrancyGuard, Ownable2StepUpgradeable {
```
Comment (kelemeno) [inline l1-contracts/contracts/transactionFilterer/TransactionFilterer.sol:15]: Can you name it GatewayTransactionFilterer?

File: l1-contracts/test/foundry/unit/concrete/TransactionFilterer/CheckTransaction.sol [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.24;
+
+import {TransactionFiltererTest} from "./_TransactionFilterer_Shared.t.sol";
+
+import {AlreadyWhitelisted, NotWhitelisted} from "contracts/common/L1ContractErrors.sol";
+
+contract checkTransactionTest is TransactionFiltererTest {
```
Comment (StanislavBreadless) [inline l1-contracts/test/foundry/unit/concrete/TransactionFilterer/CheckTransaction.sol:8]: checkTransactionTest should start with capital letter

File: l1-contracts/test/unit_tests/transaction_filterer.spec.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
+import { expect } from "chai";
```
Comment (StanislavBreadless) [inline l1-contracts/test/unit_tests/transaction_filterer.spec.ts:1]: While additional tests are always appreciated, considering that we'll migrate from hardhat to foundry soon, the hardhat tests will be eventually deleted. Is there something they test that the foundry tests don't?

Comment (Raid5594) [reply l1-contracts/test/unit_tests/transaction_filterer.spec.ts:1]: I think it's just the difference in ownership tests for coverage, but I could do the same in forge, so no real benefit, so deleting it

File: l1-contracts/test/foundry/unit/concrete/GatewayTransactionFilterer/CheckTransaction.sol [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import {IGetters} from "contracts/state-transition/chain-interfaces/IGetters.sol";
+import {IL2Bridge} from "contracts/bridge/interfaces/IL2Bridge.sol";
+import {IBridgehub} from "contracts/bridgehub/IBridgehub.sol";
+import {AlreadyWhitelisted, NotWhitelisted} from "contracts/common/L1ContractErrors.sol";
+
+contract CheckTransactionTest is GatewayTransactionFiltererTest {
+    function test_TransactionAllowedOnlyFromWhitelistedSenderWhichIsNotBaseTokenBridge() public {
+        address baseTokenBridge = address(0x4040404);
+        address bridgehub = address(0x5050505);
+        // address stm = address(0x6060606);
```
Comment (StanislavBreadless) [inline l1-contracts/test/foundry/unit/concrete/GatewayTransactionFilterer/CheckTransaction.sol:15]: ?

File: l1-contracts/contracts/transactionFilterer/GatewayTransactionFilterer.sol [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+/// @dev Only allows whitelisted senders to deposit to Gateway
+contract GatewayTransactionFilterer is ITransactionFilterer, ReentrancyGuard, Ownable2StepUpgradeable {
+    /// @dev Event emitted when sender is whitelisted
+    event WhitelistGranted(address indexed sender);
+
+    /// @dev Event emitted when sender is removed from whitelist
+    event WhitelistRevoked(address indexed sender);
+
+    /// @dev Bridgehub is set during construction
+    IBridgehub public immutable bridgeHub;
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/transactionFilterer/GatewayTransactionFilterer.sol:25]: by convention our immutable variables must be in CAPITALIZED_SNAKE_CASE

**Review by StanislavBreadless — APPROVED** [review]

**Pull request merged**